api/bucketsd/service.go

```diff
@@ -961,6 +971,247 @@ func (s *Service) createLinks(
+		return nil, fmt.Errorf("verifying bucket update: %v", err)
+	}
+
+	fromNodePath, err := inflateFilePath(buck, fromPth)
+	if err != nil {
+		return nil, err
+	}
+	fromItem, err := s.pathToItem(ctx, buck, fromNodePath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	toNodePath, err := inflateFilePath(buck, toPth)
+	if err != nil {
+		return nil, err
+	}
+
+	toItem, err := s.pathToItem(ctx, buck, toNodePath, false)
+	if err == nil {
+		if fromItem.IsDir && !toItem.IsDir {
+			return nil, fmt.Errorf("not a directory")
+		}
+		if toItem.IsDir {
+			// from => to becomes cleanDst:
+			// "c" => "b" becomes "b/c"
+			// "a.jpg" => "b" becomes "b/a.jpg"
+			toPth = gopath.Join(toPth, fromItem.Name)
+		}
+	}
+
+	// bucket node
+	buckNode, err := s.getNodeAtPath(ctx, pth, buck.GetLinkEncryptionKey())
+	if err != nil {
+		return nil, fmt.Errorf("getting node: %v", err)
+	}
+
+	var dirPath path.Resolved
+	// Private buckets need to manage keys+encrypted nodes
+	if buck.IsPrivate() {
+		ctx, dirPath, err = s.copyNode(ctx, buck, buckNode, fromPth, toPth)
+		if err != nil {
+			return nil, fmt.Errorf("copying node: %v", err)
+		}
+	} else {
+		buckPath := path.New(buck.Path)
+		ctx, dirPath, err = s.setPathFromExistingCid(
+			ctx,
+			buck,
+			buckPath,
+			toPth,
+			buckNode.Cid(),
+			nil,
+			nil,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("copying path: %v", err)
+		}
+	}
+	buck.Path = dirPath.String()
+
+	// If "a/b" => "a/", cleanup only needed for priv
+	if strings.HasPrefix(fromPth, toPth) {
+		if buck.IsPrivate() {
+			ctx, dirPath, err = s.removePath(ctx, buck, fromPth)
+			if err != nil {
+				return nil, fmt.Errorf("cleaning path: %v", err)
+			}
+			buck.Path = dirPath.String()
+		}
+
+		if err = s.saveAndPublish(ctx, dbID, dbToken, buck); err != nil {
+			return nil, err
+		}
+		return &pb.MovePathResponse{}, nil
+	}
+
+	if strings.HasPrefix(toPth, fromPth) {
+		// If "a/" => "a/b" cleanup each leaf in "a" that isn't "b" (skipping .textileseed)
+		ppth := path.Join(path.New(buck.Path), fromPth)
+		item, err := s.listPath(ctx, dbID, buck, ppth)
+		if err != nil {
+			return nil, fmt.Errorf("listing path: %v", err)
+		}
+		reg := regexp.MustCompile("/ipfs/([^/]+)/")
+		for _, chld := range item.Item.Items {
+			sp := cleanPath(reg.ReplaceAllString(chld.Path, ""))
+			if strings.Compare(chld.Name, buckets.SeedName) == 0 || sp == toPth {
+				continue
+			}
+			spClean := cleanPath(sp)
+			ctx, dirPath, err = s.removePath(ctx, buck, spClean)
+			if err != nil {
+				return nil, fmt.Errorf("cleaning path: %v", err)
+			}
+			buck.Path = dirPath.String()
+		}
+	} else {
+		// if a/ => b/ remove a
+		ctx, dirPath, err = s.removePath(ctx, buck, fromPth)
+		if err != nil {
+			return nil, fmt.Errorf("removing path: %v", err)
+		}
+		buck.Path = dirPath.String()
+	}
+
+	if err = s.saveAndPublish(ctx, dbID, dbToken, buck); err != nil {
+		return nil, err
+	}
+
+	return &pb.MovePathResponse{}, nil
+}
+
+func (s *Service) copyNode(
```

Comment:
Moved below where it's used... helps with organization a little bit in this giant file.